# Тиждень 2 - Дослідження можливостей ШІ для оптимізації коду та покращення стилю


## Puzzle (Процедурне програмування: тема 8)

#### Власний код

In [ ]:
''' validate_board(board)

githb: https://github.com/MarkusViktoria/Markus-Viktoria-lab8-task2.git '''

import psutil
import time

def validate_rows(board: list) -> bool:
    '''
    Return result of checking rows.
    >>> validate_rows(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    True
        '''
    if not isinstance(board, list):
        return None
    for rows in board:
        for elem in rows:
            if rows.count(elem) != 1 and elem.isdigit():
                return False
    return True

def validate_columns(board: list) -> bool:
    '''
    Return result of checking rows.
    >>> validate_columns(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    False
        '''
    if not isinstance(board, list):
        return None
    new_board = []
    for i in range(9):
        new_elem = ''
        for row in board:
            new_elem += row[i]
        new_board.append(new_elem)
    return validate_rows(new_board)

def validate_colors(board: list) -> bool:
    '''
    Return result of checking colored blocks.
    >>> validate_colors(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    True
    '''
    if not isinstance(board, list):
        return None
    new_board = []
    for i in range(9):
        new_elem = ''
        for row in range(9-i):
            new_elem += board[row][i]
        new_elem += board[-(i+1)][i+1:]
        new_board.append(new_elem)
    return validate_rows(new_board)

def validate_board(board: list) -> bool:
    '''
    Return result of checking board.
    >>> validate_board(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    False
    '''
    if not isinstance(board, list):
        return None
    return validate_rows(board) and validate_columns(board) and validate_colors(board)

def memory_usage():
    process = psutil.Process()
    mem = process.memory_info().rss / (1024 * 1024)
    return mem

if __name__ == '__main__':
    print(memory_usage)
    start_time = time.time()
    import doctest
    print(doctest.testmod())
    end_time = time.time()
    print("Час виконання:", end_time - start_time, "секунд")


#### Результат перевірки на об'єм пам'яті та кількість часу:

<function memory_usage at 0x0000021C2FDB79C0>

TestResults(failed=0, attempted=4)

Час виконання: 0.007683992385864258 секунд

#### ChatGPT при наданні власного коду та формулювання промпту українською мовою

1) Промт: "Оптимізуй наданий код"

In [2]:
# РЕЗУЛЬТАТ

def is_valid(board: list) -> bool:
    """
    Return True if the Sudoku board is valid, False otherwise.
    """
    if not isinstance(board, list) or len(board) != 9:
        return False

    # Check rows, columns, and colors
    return check_rows(board) and check_columns(board) and check_colors(board)


def check_rows(board: list) -> bool:
    """
    Check if each row of the board is valid.
    """
    for row in board:
        if not is_valid_unit(row):
            return False
    return True


def check_columns(board: list) -> bool:
    """
    Check if each column of the board is valid.
    """
    for col in range(9):
        column = [board[row][col] for row in range(9)]
        if not is_valid_unit(column):
            return False
    return True


def check_colors(board: list) -> bool:
    """
    Check if each color block of the board is valid.
    """
    for start_row in range(0, 9, 3):
        for start_col in range(0, 9, 3):
            block = [board[i][j] for i in range(start_row, start_row + 3) for j in range(start_col, start_col + 3)]
            if not is_valid_unit(block):
                return False
    return True


def is_valid_unit(unit: list) -> bool:
    """
    Check if a row, column, or color block is valid.
    """
    unit = [x for x in unit if x != ' ']
    return len(set(unit)) == len(unit)


if __name__ == '__main__':
    import doctest
    print(doctest.testmod())


TestResults(failed=0, attempted=4)


**Висновок:**

Як результат, згенерований код не проходить тести, тому що ШІ змінив повністю останню функцію, тобто функції, яка все умовно збирала до купи, не існує, і вся робота анульовується. Також ChatGPT видалив всі доктести і залишив тільки докстрінги.

2) Промт: "Не змінюй назви та концепції функцій, оптимізуй код,  скоротивши кількість циклів та розгалужень. Покращ стилізацію коду відповідно до pep8"

In [ ]:
def validate_rows(board: list) -> bool:
    """
    Return True if each row of the Sudoku board is valid, False otherwise.
    """
    if not isinstance(board, list):
        return None
    
    for row in board:
        if not is_valid_unit(row):
            return False
    return True


def validate_columns(board: list) -> bool:
    """
    Return True if each column of the Sudoku board is valid, False otherwise.
    """
    if not isinstance(board, list):
        return None
    
    return validate_rows([ ''.join([board[row][col] for row in range(9)]) for col in range(9)])


def validate_colors(board: list) -> bool:
    """
    Return True if each color block of the Sudoku board is valid, False otherwise.
    """
    if not isinstance(board, list):
        return None
    
    return validate_rows([
        ''.join([board[row + col][col] for col in range(9 - row)]) + ''.join([board[row + 8 - col][col + 1] for col in range(row)]) 
        for row in range(9)
    ])


def validate_board(board: list) -> bool:
    """
    Return True if the Sudoku board is valid, False otherwise.
    """
    if not isinstance(board, list):
        return None
    
    return validate_rows(board) and validate_columns(board) and validate_colors(board)


def is_valid_unit(unit: list) -> bool:
    """
    Check if a row, column, or color block is valid.
    """
    unit = [x for x in unit if x != ' ']
    return len(set(unit)) == len(unit)


if __name__ == '__main__':
    import doctest
    print(doctest.testmod())


**Висновок** 

Згенерований код, проходить тести на 0.26/1, загалом не проходять 3 тести та тестування документації. Як ми можемо бачити доктести не з'явилися, загалом зменшився обсяг самого коду.

3) Промт: "Не змінюючи назв та концепції самих функцій, оптимізуй код, ліквідувавши непотрібні цикли та розгалуження. Відредагуй доктести та докстрінги відповідно до правил pep8 та pylint. Також не додавай жодних додаткових функцій. Працюй тільки з цим кодом"

In [ ]:
'''Puzzle'''

def validate_rows(board: list) -> bool:
    """
    Check if all rows in the board are valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if all rows are valid, False otherwise
    """
    if not isinstance(board, list):
        return None
    for row in board:
        row_digits = [digit for digit in row if digit.isdigit()]
        if len(set(row_digits)) != len(row_digits):
            return False
    return True

def validate_columns(board: list) -> bool:
    """
    Check if all columns in the board are valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if all columns are valid, False otherwise
    """
    if not isinstance(board, list):
        return None
    transposed_board = [''.join(col) for col in zip(*board)]
    return validate_rows(transposed_board)

def validate_colors(board: list) -> bool:
    """
    Check if all colored blocks in the board are valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if all colored blocks are valid, False otherwise
    """
    if not isinstance(board, list):
        return None
    new_board = []
    for i in range(9):
        new_elem = ''
        for row in range(9 - i):
            new_elem += board[row][i]
        new_elem += board[-(i + 1)][i + 1:]
        new_board.append(new_elem)
    return validate_rows(new_board)

def validate_board(board: list) -> bool:
    """
    Check if the Sudoku board is valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if the board is valid, False otherwise
    """
    if not isinstance(board, list):
        return None
    return validate_rows(board) and validate_columns(board) and validate_colors(board)

if __name__ == '__main__':
    import doctest
    print(doctest.testmod())


**Висновок**

При введенні конкретного запиту з вказанням усіх побажань щодо поправок ChatGPT згенерував відносно правильний код, який проходить всі тести,які стосуються саме функцій, проте частина функцій не змінилася, тобто з оптимізацією ШІ впорався відносно погано. Також ШІ не згенерував доктести, через що на CMS завдання отримало тільки 0.5 бала. Це означає, що з стилізацією коду в ШІ також виникають проблеми.

##### Перевірка на кількість об'му пам'яті та часу реалізації.


In [ ]:
'''Puzzle'''

import psutil
import time

def validate_rows(board: list) -> bool:
    """
    Check if all rows in the board are valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if all rows are valid, False otherwise
    >>> validate_rows(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    True
    """
    if not isinstance(board, list):
        return None
    for row in board:
        row_digits = [digit for digit in row if digit.isdigit()]
        if len(set(row_digits)) != len(row_digits):
            return False
    return True

def validate_columns(board: list) -> bool:
    """
    Check if all columns in the board are valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if all columns are valid, False otherwise
    >>> validate_columns(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    False
    """
    if not isinstance(board, list):
        return None
    transposed_board = [''.join(col) for col in zip(*board)]
    return validate_rows(transposed_board)

def validate_colors(board: list) -> bool:
    """
    Check if all colored blocks in the board are valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if all colored blocks are valid, False otherwise
    >>> validate_colors(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    True
    """
    if not isinstance(board, list):
        return None
    new_board = []
    for i in range(9):
        new_elem = ''
        for row in range(9 - i):
            new_elem += board[row][i]
        new_elem += board[-(i + 1)][i + 1:]
        new_board.append(new_elem)
    return validate_rows(new_board)

def validate_board(board: list) -> bool:
    """
    Check if the Sudoku board is valid.

    :param board: Sudoku board represented as a list of strings
    :return: True if the board is valid, False otherwise
    >>> validate_board(["**** ****", "***1 ****", "**  3****", "* 4 1****",\
 "     9 5 ", " 6  83  *", "3   1  **", "  8  2***", "  2  ****"])
    False
    """
    if not isinstance(board, list):
        return None
    return validate_rows(board) and validate_columns(board) and validate_colors(board)

def memory_usage():
    process = psutil.Process()
    mem = process.memory_info().rss / (1024 * 1024)
    return mem

if __name__ == '__main__':
    print(memory_usage)
    start_time = time.time()
    import doctest
    print(doctest.testmod())
    end_time = time.time()
    print("Час виконання:", end_time - start_time, "секунд")

#### Результат:

<function memory_usage at 0x0000021C2FDB6FC0>

TestResults(failed=0, attempted=4)

Час виконання: 0.013434648513793945 секунд

### Висновок:

Порівнюючи результат перевірки на кількість витраченого об'єму пам'яті та кількості часу реалізації власного коду та коду "оптимізованого" з погляду ChatGPT, можна виявити, що оптимізований код виконується довше, тобто можна зробити висновок, що використовувати ШІ для отпимізації та стилізації коду не варто.

## Прибульці та IQ (Процедурне програмування: тема 11)

#### Власний код

In [15]:
'''Alien'''

import psutil
import time

def read_file(file_path):
    '''
    Read file and return dictionary of people and their IQ level.
    >>> read_file('./smart_people.txt')
    {'Elon Musk': 165, 'Mark Zuckerberg': 152, 'Will Smith': 157, 'Marilyn vos Savant': 186,\
 'Judith Polgar': 170, 'Quentin Tarantino': 163, 'Bill Gates': 160, "Conan O'Brien": 160,\
 'Emma Watson': 132, 'Barack Obama': 137}
    '''
    with open(file_path, 'r', encoding='utf-8') as file:
        data = file.readlines()
    smarties = {}
    for line in data[1:]:
        line = line.split(',')
        smarties[line[0]] = int(line[-1])
    return smarties

def rescue_people(smarties, limit_iq):
    '''
    Return list of trip and people.
    >>> rescue_people({"Steve Jobs": 160, "Albert Einstein": 160, "Sir Isaac Newton": 195,\
 "Nikola Tesla": 189},500)
    (2, [['Sir Isaac Newton', 'Nikola Tesla'], ['Albert Einstein', 'Steve Jobs']])
    >>> rescue_people({"Alice": 100, "Bob": 80, "Charlie": 90}, 180)
    (2, [['Alice', 'Bob'], ['Charlie']])
    >>> rescue_people({"Alice": 100, "Bob": 80, "Charlie": 90}, 110)
    (3, [['Alice'], ['Charlie'], ['Bob']])
    '''
    result = []

    if smarties == {} or limit_iq == 0:
        return (0, [])

    lst_of_people = sorted(smarties.items(), key=lambda x: (-x[1], x[0]))

    while lst_of_people:
        temp_res = []
        temp_limit = limit_iq

        for elem in lst_of_people:
            if elem[1] <= temp_limit:
                temp_res.append(elem[0])
                temp_limit -= elem[1]

        result.append(temp_res)
        lst_of_people = [e for e in lst_of_people if e[0] not in temp_res]

    return (len(result), result)

def memory_usage():
    process = psutil.Process()
    mem = process.memory_info().rss / (1024 * 1024)
    return mem

if __name__ == '__main__':
    print(memory_usage)
    start_time = time.time()
    import doctest
    print(doctest.testmod())
    end_time = time.time()
    print("Час виконання:", end_time - start_time, "секунд")


<function memory_usage at 0x0000021C2FDB7600>
TestResults(failed=0, attempted=8)
Час виконання: 0.00996851921081543 секунд


#### Результат перевірки на об'єм пам'яті та кількість часу:

<function memory_usage at 0x0000021C2FDB7600>

TestResults(failed=0, attempted=8)

Час виконання: 0.00996851921081543 секунд

#### ChatGPT при наданні власного коду та формулювання промпту українською мовою

1) Промт: "Оптимізуй наданий код"

In [ ]:
from collections import defaultdict

def read_file(file_path):
    '''
    Read file and return dictionary of people and their IQ level.
    >>> read_file('./smart_people.txt')
    {'Elon Musk': 165, 'Mark Zuckerberg': 152, 'Will Smith': 157, 'Marilyn vos Savant': 186,\
 'Judith Polgar': 170, 'Quentin Tarantino': 163, 'Bill Gates': 160, "Conan O'Brien": 160,\
 'Emma Watson': 132, 'Barack Obama': 137}
    '''
    smarties = defaultdict(int)
    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file.readlines()[1:]:
            name, iq = line.strip().split(',')
            smarties[name] = int(iq)
    return smarties

def rescue_people(smarties, limit_iq):
    '''
    Return list of trip and people.
    >>> rescue_people({"Steve Jobs": 160, "Albert Einstein": 160, "Sir Isaac Newton": 195,\
 "Nikola Tesla": 189},500)
    (2, [['Sir Isaac Newton', 'Nikola Tesla'], ['Albert Einstein', 'Steve Jobs']])
    >>> rescue_people({"Alice": 100, "Bob": 80, "Charlie": 90}, 180)
    (2, [['Alice', 'Bob'], ['Charlie']])
    >>> rescue_people({"Alice": 100, "Bob": 80, "Charlie": 90}, 110)
    (3, [['Alice'], ['Charlie'], ['Bob']])
    '''
    result = []

    if not smarties or limit_iq == 0:
        return (0, [])

    lst_of_people = sorted(smarties.items(), key=lambda x: (-x[1], x[0]))

    while lst_of_people:
        temp_res = []
        temp_limit = limit_iq

        for elem in lst_of_people:
            if elem[1] <= temp_limit:
                temp_res.append(elem[0])
                temp_limit -= elem[1]

        result.append(temp_res)
        lst_of_people = [e for e in lst_of_people if e[0] not in temp_res]

    return (len(result), result)

if __name__ == '__main__':
    import doctest
    print(doctest.testmod())


**Висновок**

На диво ChatGPT згенерував відносно правильний код, навіть при такому короткому запиті. Тести пов'язані саме з кодом, а не з документацією згенерований код пройшов і отримав 0.5 балів з 1 на CMS. Проте ще варто попрацювати над документацією.

2) Промт: "Не змінюючи назв та концепції самих функцій, оптимізуй код, ліквідувавши непотрібні цикли та розгалуження. Відредагуй доктести та докстрінги відповідно до правил pep8 та pylint. Також не додавай жодних додаткових функцій. Працюй тільки з цим кодом"

In [ ]:
def read_file(file_path):
    """
    Read file and return dictionary of people and their IQ level.
    
    :param file_path: Path to the file
    :type file_path: str
    :return: Dictionary containing people and their IQ levels
    :rtype: dict
    """
    smarties = {}
    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file.readlines()[1:]:
            name, iq = line.strip().split(',')
            smarties[name] = int(iq)
    return smarties

def rescue_people(smarties, limit_iq):
    """
    Return list of trips and people.
    
    :param smarties: Dictionary containing people and their IQ levels
    :type smarties: dict
    :param limit_iq: Limit of IQ for the rescue trip
    :type limit_iq: int
    :return: Tuple containing the number of trips and list of people in each trip
    :rtype: tuple
    """
    if not smarties or limit_iq == 0:
        return (0, [])

    lst_of_people = sorted(smarties.items(), key=lambda x: (-x[1], x[0]))
    result = []

    while lst_of_people:
        temp_res = []
        temp_limit = limit_iq

        for elem in lst_of_people[:]:
            if elem[1] <= temp_limit:
                temp_res.append(elem[0])
                temp_limit -= elem[1]
                lst_of_people.remove(elem)

        result.append(temp_res)

    return (len(result), result)

if __name__ == '__main__':
    import doctest
    print(doctest.testmod())


**Висновок**

Цього разу чітко прописаний промт зробив тільки гірше, і балів стало на 0.01 менше, знову ж таки через тестування документації. Це вкотре підтверджує проблему стилізації коду в ШІ.

##### Перевірка на кількість об'му пам'яті та часу реалізації.


In [17]:
'''Alien'''

import psutil
import time


def read_file(file_path):
    """
    Read file and return dictionary of people and their IQ level.
    
    :param file_path: Path to the file
    :type file_path: str
    :return: Dictionary containing people and their IQ levels
    :rtype: dict
    >>> read_file('./smart_people.txt')
    {'Elon Musk': 165, 'Mark Zuckerberg': 152, 'Will Smith': 157, 'Marilyn vos Savant': 186,\
 'Judith Polgar': 170, 'Quentin Tarantino': 163, 'Bill Gates': 160, "Conan O'Brien": 160,\
 'Emma Watson': 132, 'Barack Obama': 137}
    """
    smarties = {}
    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file.readlines()[1:]:
            name, iq = line.strip().split(',')
            smarties[name] = int(iq)
    return smarties

def rescue_people(smarties, limit_iq):
    """
    Return list of trips and people.
    
    :param smarties: Dictionary containing people and their IQ levels
    :type smarties: dict
    :param limit_iq: Limit of IQ for the rescue trip
    :type limit_iq: int
    :return: Tuple containing the number of trips and list of people in each trip
    :rtype: tuple
    >>> rescue_people({"Steve Jobs": 160, "Albert Einstein": 160, "Sir Isaac Newton": 195,\
 "Nikola Tesla": 189},500)
    (2, [['Sir Isaac Newton', 'Nikola Tesla'], ['Albert Einstein', 'Steve Jobs']])
    >>> rescue_people({"Alice": 100, "Bob": 80, "Charlie": 90}, 180)
    (2, [['Alice', 'Bob'], ['Charlie']])
    >>> rescue_people({"Alice": 100, "Bob": 80, "Charlie": 90}, 110)
    (3, [['Alice'], ['Charlie'], ['Bob']])
    """
    if not smarties or limit_iq == 0:
        return (0, [])

    lst_of_people = sorted(smarties.items(), key=lambda x: (-x[1], x[0]))
    result = []

    while lst_of_people:
        temp_res = []
        temp_limit = limit_iq

        for elem in lst_of_people[:]:
            if elem[1] <= temp_limit:
                temp_res.append(elem[0])
                temp_limit -= elem[1]
                lst_of_people.remove(elem)

        result.append(temp_res)

    return (len(result), result)

def memory_usage():
    process = psutil.Process()
    mem = process.memory_info().rss / (1024 * 1024)
    return mem

if __name__ == '__main__':
    print(memory_usage)
    start_time = time.time()
    import doctest
    print(doctest.testmod())
    end_time = time.time()
    print("Час виконання:", end_time - start_time, "секунд")


<function memory_usage at 0x0000021C302BA200>
TestResults(failed=0, attempted=8)
Час виконання: 0.012958288192749023 секунд


#### Результат перевірки на об'єм пам'яті та кількість часу:

<function memory_usage at 0x0000021C302BA200>

TestResults(failed=0, attempted=8)

Час виконання: 0.012958288192749023 секунд

### Висновок:

Порівнюючи результат перевірки на кількість витраченого об'єму пам'яті та кількості часу реалізації власного коду та коду "оптимізованого" з погляду ChatGPT,  знову випливає те, що оптимізований код виконується довше, що знову ж таки вказує на те, що не варто використовувати ШІ для оптимізаційних процесів.

# Підсумок:

Використання ШІ та зокрема ChatGPT для оптимізації та стилізації коду є марною тратою часу, оскільки загальна картина в кращому випадку не зміниться, а в гіршому код буде реалізовуватися довше. Можна було помітити, що окремі невеликі функції ChatGPT загалом може оптимізувати, наприклад прибравши кілька циклів і записати це все у list comperhension. Проте нажаль повну реалізацію задачі оптимізувати йому вдається в край важко, хоча з конкретно прописаним промтом це легше зробити.

Щодо стилізації коду та написання документації, то тут все дуже сумно. Навіть даючи посилання на pep8 та pylint ChatGPT не спромігся нормально написати документацію, навіть даючи завтання - написати доктести та докстрінги. 

Тож основним висновком є те, що не варто покладатися на ШІ в питаннях оптимізації та стилізації коду